### PR TITLE
Use nocloud when specified

### PR DESCRIPTION
**Pull request by hh**
We are picking this up from @BobyMCbobs  (thanks for the work on this)
Context: https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/212

### CODE REVIEW

**Comment by hh** [comment]
Ensuring we loop in @davidvossel 

> someone could theoretically set a CloudInitConfigDrive volume too. I don't like that this logic is assuming that only a CloudInitNoCloud volume get's the special behavior.

> making the default cloudInitConfigDrive is good though

> someone could theoretically set a CloudInitConfigDrive volume too. I don't like that this logic is assuming that only a CloudInitNoCloud volume get's the special behavior.

> making the default cloudInitConfigDrive is good though

**Comment by davidvossel** [comment]
this comment [1] from the previous pr was the primary remaining thing. We don't have to remove mergo from go.mod since i know it's an indirect dependency somewhere in the dependency chain. I'd just like to see that we dont' introduce that dependency directly in the capk code for this PR simply because it isn't needed.

1. https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/212#discussion_r1047755256

File: pkg/kubevirt/utils.go [hunk 1]
```diff
@@ -156,17 +155,10 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) (*kubevirt
...
-		}
-		return template, nil
-	case cloudInitConfigDrive:
-		template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
-	}
-	if !detectCloudInitDisk(template) {
+
+	if index >= 0 {
+		template.Spec.Volumes[index] = cloudInitVolume
+	} else {
```
Comment (hh) [inline pkg/kubevirt/utils.go:161]: Like this @davidvossel ?

File: go.mod [hunk 2]
```diff
@@ -61,7 +62,6 @@ require (
 	github.com/googleapis/gnostic v0.5.5 // indirect
 	github.com/gorilla/websocket v1.5.0 // indirect
 	github.com/gregjones/httpcache v0.0.0-20181110185634-c63ab54fda8f // indirect
-	github.com/imdario/mergo v0.3.12 // indirect
```
Comment (hh) [inline go.mod:64]: Removing mergo as requested

File: pkg/kubevirt/utils.go [hunk 3]
```diff
@@ -138,36 +152,70 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
...
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+	cloudInitType, index := detectCloudInitVolumeType(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template)
+
+	if index >= 0 {
+		template.Spec.Volumes[index] = cloudInitVolume
+	} else {
+		template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
+	}
```
Comment (hh) [inline pkg/kubevirt/utils.go:163]: Is this what you are looking for @davidvossel ?

File: pkg/kubevirt/utils.go [hunk 4]
```diff
@@ -219,12 +219,3 @@ func detectCloudInitVolumeType(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpe
...
 }
-
-func detectCloudInitDisk(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec) (foundCloudInitDisk bool) {
-	for _, v := range vmi.Spec.Domain.Devices.Disks {
-		if v.Disk != nil && v.Name == cloudInitVolumeName {
-			return true
-		}
-	}
-	return false
-}
```
Comment (hh) [inline pkg/kubevirt/utils.go:173]: I'm not sure it makes sense to detect if it's a CloudInitDisk or not anymore?

File: pkg/kubevirt/utils.go [hunk 5]
```diff
@@ -156,17 +155,10 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) (*kubevirt
...
-	case cloudInitNoCloud:
-		err := mergo.Merge(&template.Spec.Volumes[index], cloudInitVolume)
-		if err != nil {
-			return nil, err
-		}
-		return template, nil
-	case cloudInitConfigDrive:
-		template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
-	}
-	if !detectCloudInitDisk(template) {
```
Comment (hh) [inline pkg/kubevirt/utils.go]: Just making sure it makes sense to remove this switch & merge logic with the simplified one below:

File: pkg/kubevirt/utils.go (hunk 3 above)
Comment (davidvossel) [reply pkg/kubevirt/utils.go:163]: not exactly.

The disk and volume logic needs to be independent. Here's what we need.

If the volume already exists, override it `template.Spec.Volumes[index] = cloudInitVolume` else append the volume to the template.Spec.Volumes slice.

If the disk matching the volume name already exists, then leave it alone. If the disk doesn't exist, append it.

The idea here is that someone can create their own disk mapping to the cloud init volume, using whatever parameters they want and we'll use that... but if no disk is provided that matches the cloud-init volume name, we'll automatically create one using the virtio bus.

And for the volumes, some one can create a volume with a specific cloud init type, and we'll inject the user data into that, or if no volume exists already for cloud init, we'll create one for the user defaulting to cloud init config drive.

**Review by davidvossel — COMMENTED** [review]
i left a comment inline, before this merges we'll need to squash the commits down

**Pull request closed without merging**